File: algorithms/easy/number_of_valid_clock_times.ex
defmodule Solution do
  @spec count_time(time :: String.t()) :: integer
  def count_time(time) do
    [hh, mm] = String.split(time, ":") |> Enum.map(&String.to_charlist/1)
    count_hh(hh) * count_mm(mm)
  end

  def count_hh([??, ??]), do: 24

  def count_hh([??, h]) do
    if h in ?0..?3, do: 3, else: 2
  end

  def count_hh([?2, ??]), do: 4
  def count_hh([_, ??]), do: 10
  def count_hh(_), do: 1

  def count_mm([??, ??]), do: 60
  def count_mm([??, _]), do: 6
  def count_mm([_, ??]), do: 10
  def count_mm(_), do: 1
end

File: algorithms/easy/number_of_valid_clock_times.cpp
class Solution {
public:
  int countTime(string time) {
    int ct = 1;
    if (time[0] == '?') {
      ct = time[1] == '?' ? 24 : (time[1] < '4' ? 3 : 2);
    } else if (time[1] == '?') {
      ct = time[0] == '2' ? 4 : 10;
    }

    if (time[3] == '?') ct *= 6;
    if (time[4] == '?') ct *= 10;
    
    return ct;
  }
};
